ddtrace/internal/_encoding.pyx

```diff
@@ -91,180 +91,215 @@ cdef class Packer(object):
+            ret = msgpack_pack_nil(&self.pk)
+
+        if PyDict_CheckExact(meta):
+            d = <dict> meta
+            L = len(d)
+            if L > ITEM_LIMIT:
+                raise ValueError("dict is too large")
+
+            ret = msgpack_pack_map(&self.pk, L)
+            if ret == 0:
+                for k, v in d.items():
+                    ret = self._pack_text(k)
+                    if ret != 0: break
+                    ret = self._pack_text(v)
+                    if ret != 0: break
+            return ret
 
-                    ret = pack_bytes(&self.pk, <char *>b"span_id", 7)
-                    if ret != 0: return ret
-                    ret = self._pack(o.span_id)
-                    if ret != 0: return ret
+        raise TypeError("Unhandled meta type: %r" % type(meta))
 
-                    ret = pack_bytes(&self.pk, <char *>b"service", 7)
-                    if ret != 0: return ret
-                    ret = self._pack(o.service)
-                    if ret != 0: return ret
+    cdef inline int _pack_metrics(self, object metrics):
+        cdef Py_ssize_t L
+        cdef int ret
+        cdef dict d
 
-                    ret = pack_bytes(&self.pk, <char *>b"resource", 8)
-                    if ret != 0: return ret
-                    ret = self._pack(o.resource)
-                    if ret != 0: return ret
+        if metrics is None:
+            ret = msgpack_pack_nil(&self.pk)
+
+        if PyDict_CheckExact(metrics):
+            d = <dict> metrics
+            L = len(d)
+            if L > ITEM_LIMIT:
+                raise ValueError("dict is too large")
+
+            ret = msgpack_pack_map(&self.pk, L)
+            if ret == 0:
+                for k, v in d.items():
+                    ret = self._pack_text(k)
+                    if ret != 0: break
+                    ret = self._pack_number(v)
+                    if ret != 0: break
+            return ret
 
-                    ret = pack_bytes(&self.pk, <char *>b"name", 4)
-                    if ret != 0: return ret
-                    ret = self._pack(o.name)
-                    if ret != 0: return ret
+        raise TypeError("Unhandled metrics type: %r" % type(metrics))
 
-                    ret = pack_bytes(&self.pk, <char *>b"error", 5)
+    cdef int _pack(self, object o) except -1:
+        cdef int ret
+        cdef Py_ssize_t L
+        cdef int has_span_type
+        cdef int has_meta
+        cdef int has_metrics
+
+        if o is None:
+            ret = msgpack_pack_nil(&self.pk)
+
+        elif PyList_CheckExact(o):
+            # Expect a list of traces or a list of spans
+            L = len(o)
+            if L > ITEM_LIMIT:
+                raise ValueError("list is too large")
+
+            ret = msgpack_pack_array(&self.pk, L)
+            if ret != 0: return ret
+
+            for e in o:
+                ret = self._pack(e)
+                if ret != 0: break
+
+        elif isinstance(o, Span):
```

Comment:
should we move this into a `self._pack_span(o)`?

Reply:
Good idea, even better 😄